node/src/components/network.rs

```diff
@@ -599,24 +633,66 @@ where
+                        .validator_matrix
+                        .is_active_or_upcoming_validator(public_key);
+
+                    // Find the shared `Arc` that holds the validator status for this specific key.
+                    match self.incoming_validator_status.entry((**public_key).clone()) {
+                        // TODO: Use `Arc` for public key-key.
+                        Entry::Occupied(mut occupied) => {
+                            match occupied.get().upgrade() {
+                                Some(arc) => {
+                                    arc.store(status, Ordering::Relaxed);
+                                    arc
+                                }
+                                None => {
+                                    // Failed to ugprade, the weak pointer is just a leftover that
+                                    // has not been cleaned up yet. We can replace it.
+                                    let arc = Arc::new(AtomicBool::new(status));
+                                    occupied.insert(Arc::downgrade(&arc));
+                                    arc
+                                }
+                            }
+                        }
+                        Entry::Vacant(vacant) => {
+                            let arc = Arc::new(AtomicBool::new(status));
+                            vacant.insert(Arc::downgrade(&arc));
+                            arc
+                        }
+                    }
+                });
+
+                let (read_half, write_half) = tokio::io::split(transport);
+
+                let (rpc_client, rpc_server) = self.rpc_builder.build(read_half, write_half);
+
                 // Now we can start the message reader.
                 let boxed_span = Box::new(span.clone());
                 effects.extend(
-                    tasks::message_reader(
+                    tasks::message_receiver(
                         self.context.clone(),
-                        stream,
-                        self.incoming_limiter
-                            .create_handle(peer_id, peer_consensus_public_key),
-                        self.channel_management().close_incoming_receiver.clone(),
+                        validator_status,
+                        rpc_server,
+                        self.shutdown_fuse.inner().clone(),
                         peer_id,
                         span.clone(),
                     )
                     .instrument(span)
-                    .event(move |result| Event::IncomingClosed {
-                        result,
-                        peer_id: Box::new(peer_id),
-                        peer_addr,
-                        span: boxed_span,
+                    .event(move |result| {
+                        // We keep the client around, even though we do not use it, since dropping
+                        // it will cause the connection to be closed from our end.
+                        drop(rpc_client);
```

Comment:
It's just worded weirdly. "Keep it around" here means moving it into the `.event` callback, which is only run after `task::message_receiver` has concluded, which will happen if the other peer closes the connection, or an error occurs.

As a result, `rpc_client` is only dropped when the connection has been closed anyway -- if it was dropped earlier, the actual act of doing so would cause the connection to be closed.

I'll update the wording on the comment.

Reply:
See b7f6a04b6